File: leetcode/python/LEET241.py
from typing import *

class Solution:
    def diffWaysToCompute(self, expression: str) -> List[int]:
        def compute(left, right, op):
            return [eval(str(l)+op+str(r)) for l in left for r in right]
        
        if expression.isdigit():
            return [int(expression)]

        results = []
        for index, value in enumerate(expression):
            if value in '-+*':
                left = self.diffWaysToCompute(expression[:index])
                right = self.diffWaysToCompute(expression[index+1:])

                results.extend(compute(left, right, value))

        return results

sol = Solution()
print(sol.diffWaysToCompute("2-1-1"))

File: 0_3.leetcode/java/LEET241.java
import java.util.ArrayList;
import java.util.List;

public class LEET241 {
    public static void main(String[] args) {
        System.out.println(diffWaysToCompute("11"));
    }
    static char add = '+';
    static char subs = '-';
    static char mul = '*';
    public static List<Integer> diffWaysToCompute(String expression) {
        if (expression.length() == 0) {
            return new ArrayList<>();
        }
        List<Integer> results = new ArrayList<>();
        if (isNumber(expression)) {
            results.add(Integer.valueOf(expression));
            return results;
        }
        
        for (int i=0;i<expression.length();i++) {
            char value = expression.charAt(i);

            if (value == add || value == subs || value == mul) {
                List<Integer> left = diffWaysToCompute(expression.substring(0, i));
                List<Integer> right = diffWaysToCompute(expression.substring(i+1, expression.length()));
                
                List<Integer> sub_results = compute(left, right, value);
                results.addAll(sub_results);
            }
        }
        return results;
    }

    public static boolean isNumber(String input) {
        for (char ch : input.toCharArray()) {
            if (!Character.isDigit(ch)){
                return false;
            }
        }
        return true;
    }

    public static List<Integer> compute(List<Integer> left, List<Integer> right, char op) {
        List<Integer> result = new ArrayList<>();
        for (int l : left) {
            for (int r : right) {
                if (op == subs) {
                    result.add(l - r);
                } else if (op == add) {
                    result.add(l + r);
                } else {
                    result.add(l * r);
                }
            }
        } 
        return result;
    }
}
